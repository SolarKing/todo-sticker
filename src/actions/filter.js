// @flow

export type FilterActions =
    | DueDateFilterAction;

export type DueDateFilterAction = {|
    type: 'dueDateFilter',
    filter: string,
|}

export function settingDueDateFilter(dueDateFilter: string): DueDateFilterAction {
    return {
        type: 'dueDateFilter',
        filter: dueDateFilter
    };
}
