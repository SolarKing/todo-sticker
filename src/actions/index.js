// @flow
import type { TodoActions } from './todo';
import type { FilterActions } from './filter';
import type { SortActions } from './sort';

export type Actions =
    | TodoActions
    | FilterActions
    | SortActions;
