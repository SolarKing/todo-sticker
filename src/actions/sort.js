// @flow
import type {SortByType, SortColumn} from '../constant/sort';

export type SortActions =
    | SortByAction
    | SortColumnAction;

export type SortColumnAction = {|
    type: 'sortColumn',
    column: SortColumn
|};

export type SortByAction = {|
    type: 'sortBy',
    sortBy: SortByType
|};

export function settingSortColumn(column: SortColumn): SortColumnAction {
    return {
        type: 'sortColumn',
        column: column
    };
}
