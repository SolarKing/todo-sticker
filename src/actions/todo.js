// @flow
import type { TodoState } from '../states';

export type TodoActions = AddAction
    | EditAction
    | DeleteAction
    | CompleteAction;

export const TodoActionTypeEnum = {
    Add: 'add',
    Edit: 'edit',
    Delete: 'delete',
    Complete: 'complete'
};

export type AddAction = {|
    type: 'add',
    todoState: TodoState
|}

export type EditAction = {|
    type: 'edit',
    todoState: TodoState
|}

export type DeleteAction = {|
    type: 'delete',
    id: string
|}

export type CompleteAction = {|
    type: 'complete',
    id: string
|}

export function addTodo(todoState: TodoState): AddAction {
    return {
        type: TodoActionTypeEnum.Add,
        todoState
    };
}

export function editTodo(todoState: TodoState): EditAction {
    return {
        type: TodoActionTypeEnum.Edit,
        todoState
    };
}

export function deleteTodo(id: string): DeleteAction {
    return {
        type: TodoActionTypeEnum.Delete,
        id
    };
}

export function completeTodo(id: string): CompleteAction {
    return {
        type: TodoActionTypeEnum.Complete,
        id
    };
}
