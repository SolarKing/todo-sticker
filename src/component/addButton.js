// @flow
import React from 'react';
import Translate from '../class/translate';
import AddModal from '../container/addTodoModalContainer';
import styled from 'styled-components';

type Props = {
}

type States = {
    showEditModal: boolean,
    defaultTimestamp: number
}

export default class AddButton extends React.Component<Props, States> {
    handleOpenModal: Function;
    handleCloseModal: Function;
    handleOkModal: Function;

    state = {
        showEditModal: false,
        defaultTimestamp: new Date().getTime()
    };

    constructor(props: Props) {
        super(props);
        this.handleOpenModal = this.handleOpenModal.bind(this);
        this.handleCloseModal = this.handleCloseModal.bind(this);
        this.handleOkModal = this.handleOkModal.bind(this);
    }

    handleOkModal() {
        this.setState({ showEditModal: false });
    }

    handleOpenModal() {
        this.setState({
            showEditModal: true,
            defaultTimestamp: new Date().getTime()
        });
    }

    handleCloseModal() {
        this.setState({ showEditModal: false });
    }

    render() {

        return (
            <div {...this.props}>
                <AddButtonIcon onClick={this.handleOpenModal}>
                    <span className="glyphicon glyphicon-plus-sign"></span>
                </AddButtonIcon>
                <AddModal
                    onSave={this.handleOkModal}
                    onCancel={this.handleCloseModal}
                    title={Translate.tr('Add Event')}
                    visible={this.state.showEditModal}
                    defaultDueDatetime={this.state.defaultTimestamp}
                    defaultDescription={''}
                />
            </div>
        );
    }
}

const AddButtonIcon = styled.div`
    font-size: 38px !important;
    cursor: pointer;
    
    &:hover {
        color: #6494c4;
    }
`;
AddButtonIcon.displayName = 'AddButtonIcon';