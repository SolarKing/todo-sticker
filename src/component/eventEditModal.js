// @flow
import { Modal, DatePicker, TimePicker, Input } from 'antd';
import React from 'react';
import ReactDOM from 'react-dom';
import moment from 'moment';
import Translate from '../class/translate';
import type {TodoState} from '../states';

type Props = {
    title: string,
    todoState?: TodoState,
    onSave: (newTodoState: TodoState) => void,
}

type States = {
    todoState: TodoState,
    visible: boolean,
}

export default class EditModal extends React.Component<Props, States> {

    static defaultProps = {
        visible: false,
        defaultDescriptin: '',
    }

    constructor(props: Props) {
        super(props);

        let defaultDate = new Date();
        defaultDate.setUTCSeconds(0);

        const {
            todoState = {
                dueDatetime: defaultDate.getTime(),
                description: ''
            }
        } = this.props;

        this.state = {
            todoState,
            visible: true,
        };
    }

    _handleDateChanged(dates: moment) {
        let {
            todoState
        } = this.state;

        let dueDatetime = moment(todoState.dueDatetime);
        dueDatetime.set('year', dates.get('year'));
        dueDatetime.set('month', dates.get('month'));
        dueDatetime.set('date', dates.get('date'));
        todoState.dueDatetime = dueDatetime.toDate().getTime();

        this.setState({
            todoState
        });
    }

    _handleTimeChanged(time: moment) {
        let {
            todoState
        } = this.state;

        let dueDatetime = moment(todoState.dueDatetime);
        dueDatetime.set('hour', time.get('hour'));
        dueDatetime.set('minute', time.get('minute'));
        todoState.dueDatetime = dueDatetime.toDate().getTime();

        this.setState({
            todoState
        });
    }

    _handleDescriptionChanged(event: SyntheticEvent<HTMLTextAreaElement>) {
        let {
            todoState
        } = this.state;
        todoState.description = event.currentTarget.value;

        this.setState({
            todoState
        });
    }

    _handleOk() {
        this.props.onSave(this.state.todoState);
        this.setState({
            visible: false,
        });
    }

    _handleCancel() {
        this.setState({
            visible: false,
        });
    }

    render() {
        const {
            onSave,
            title,
            // onClose,
            ...others
        } = this.props;

        const {
            todoState
        } = this.state;

        return (
            <Modal
                {...others}
                title={title}
                destroyOnClose={true}
                onOk={this._handleOk.bind(this)}
                onCancel={this._handleCancel.bind(this)}
                visible={this.state.visible}
            >
                <p>{Translate.tr('Due date')}</p>
                <DatePicker 
                    allowClear={false}
                    onChange={this._handleDateChanged.bind(this)}
                    defaultValue={moment(todoState.dueDatetime)}
                />
                &nbsp;
                <TimePicker
                    allowEmpty={false}
                    onChange={this._handleTimeChanged.bind(this)}
                    defaultValue={moment(todoState.dueDatetime)}
                    format={'HH:mm'}
                />
                <p>{Translate.tr('eventDescription')}</p>
                <Input.TextArea
                    defaultValue={todoState.description}
                    onChange={this._handleDescriptionChanged.bind(this)}
                    rows={4}
                />
            </Modal>
        );
    }
}

export const TodoEditableModal = {
    show(props: Props) {
        let wrapper = document.getElementById('todoModal');
        if (!wrapper) {
            return;
        }

        let cleanup = () => {
            ReactDOM.unmountComponentAtNode(wrapper);
        };

        ReactDOM.render(
            <EditModal
                {...props}
                afterClose={cleanup}
            />
            , wrapper
        );
    }
};