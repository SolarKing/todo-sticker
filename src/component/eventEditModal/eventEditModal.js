// @flow
import { Modal, DatePicker, TimePicker, Input } from 'antd';
import React from 'react';
import moment from 'moment';
import 'moment/locale/zh-cn';
import Translate from '../../class/translate';

export type EditInfo = {
    date: number,
    time: number,
    desctiption: string
}

type Props = {
    visible: boolean,
    title: string,
    defaultDate?: number,
    defaultTime?: number,
    defaultDescriptin?: string,
    onSave?: (editInfo: EditInfo) => void
}

type States = {
    date: number,
    time: number,
    desctiption: string
}

export default class EditModal extends React.Component<Props, States> {

    handleDateChanged: Function;
    handleTimeChanged: Function;
    handleDescriptionChanged: Function;
    handleOk: Function;

    static defaultProps = {
        visible: false,
        defaultDate: new Date().getTime(),
        defaultTime: new Date().getTime(),
        defaultDescriptin: '',
    }

    state = {
        date: new Date().getTime(),
        time: new Date().getTime(),
        desctiption: ''
    };

    constructor(props: Props) {
        super(props);
        moment.locale('en');
        this.handleDateChanged = this.handleDateChanged.bind(this);
        this.handleTimeChanged = this.handleTimeChanged.bind(this);
        this.handleDescriptionChanged = this.handleDescriptionChanged.bind(this);
        this.handleOk = this.handleOk.bind(this);
    }

    handleDateChanged(dates: moment) {
        this.setState(
            {
                date: dates
            }
        );
    }

    handleTimeChanged(time: moment) {
        this.setState(
            {
                time: time
            }
        );
    }

    handleDescriptionChanged(event: SyntheticEvent<HTMLTextAreaElement>) {
        let descriptin = event.currentTarget.value;
        this.setState({
            desctiption: descriptin
        });
    }

    handleOk() {
        if (!this.props.onSave) {
            return;
        }
        let editInfo: EditInfo;
        editInfo = {
            date: this.state.date,
            time: this.state.time,
            desctiption: this.state.desctiption
        };
        this.props.onSave(editInfo);
    }

    render() {
        var {
            defaultDate,
            defaultTime,
            defaultDescriptin,
            ...others
        } = this.props;

        return (
            <Modal onOk={this.handleOk} {...others} >
                <p>{Translate.tr('notificationDate')}</p>
                <DatePicker onChange={this.handleDateChanged} value={moment(defaultDate)} />
                <p>{Translate.tr('notificationTime')}</p>
                <TimePicker onChange={this.handleTimeChanged} value={moment(defaultTime)} format={'HH:mm'} />
                <p>{Translate.tr('eventDescription')}</p>
                <Input.TextArea defaultValue={defaultDescriptin} onChange={this.handleDescriptionChanged} rows={4} />
            </Modal>
        );
    }
}
