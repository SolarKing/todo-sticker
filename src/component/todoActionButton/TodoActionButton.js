// @flow
import React from 'react';
import { Icon } from 'antd';
import { Menu, Dropdown, Button } from 'antd';
import EditModal from '../../container/editTodoModalContainer';
import Translate from '../../class/translate';
import type {TodoState} from '../../states/index';
import DeleteButtonContainer from '../../container/DeleteButtonContainer';

type Props = {
    todo: TodoState,
}

type States = {
    editModalVisible: boolean,
}

export default class TodoActionButton extends React.Component<Props, States> {

    handleEdit: Function;

    state: States = {
        editModalVisible: false,
    }

    constructor(props: Props) {
        super(props);
        this.handleEdit = this.handleEdit.bind(this);
    }

    handleEdit() {
        this.setState({
            editModalVisible: true
        });
    }

    render() {
        let {
            todo,
            ...others
        } = this.props;

        const menu = (
            <Menu>
                <Menu.Item>
                    <a>Complete</a>
                </Menu.Item>
                <Menu.Item>
                    <a onClick={this.handleEdit} >Edit</a>
                </Menu.Item>
                <Menu.Item>
                    <DeleteButtonContainer todoId={todo.id} />
                </Menu.Item>
            </Menu>
        );
        return (
            <div {...others}>
                <Dropdown overlay={menu} trigger={['click']}>
                    <Button><Icon type="ellipsis" /></Button>
                </Dropdown>
                <EditModal
                    title={Translate.tr('Edit Todo')}
                    visible={this.state.editModalVisible}
                    onSave={() => this.setState({editModalVisible: false})}
                    onCancel={() => this.setState({editModalVisible: false})}
                    todoId={todo.id}
                    defaultDate={todo.notificationDate || new Date().getTime()}
                    defaultTime={todo.notificationTime || new Date().getTime()}
                    defaultDescription={todo.description}
                />
            </div>
        );
    }
}
