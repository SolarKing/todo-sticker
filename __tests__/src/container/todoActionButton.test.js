// @flow
import React from 'react';
import { shallow, type ShallowWrapper } from 'enzyme';
import TodoActionButton from '../../../src/container/todoActionButton';
import { Dropdown, Menu } from 'antd';
import { TodoEditableModal } from '../../../src/component/eventEditModal';
import configureStore from 'redux-mock-store';
import { editTodo, completeTodo, deleteTodo } from '../../../src/actions/todo';
import ConfirmButton from '../../../src/component/confirmButton';
import stickerDispatcher from '../../../src/sticker/dispatcher';

jest.mock('electron', () => {
    let ipcRenderer = {
        send: jest.fn()
    };
    return {
        ipcRenderer
    };
});

describe('TodoActionButton', () => {
    let wrapperContainer: ShallowWrapper;
    let wrapper: ShallowWrapper;
    let store = configureStore()();
    beforeAll(() => {
        wrapperContainer= shallow(
            <TodoActionButton
                todo={{
                    id: '1',
                    dueDatetime: 10,
                    description: 'Hello!',
                    complete: false,
                    createTime: 0,
                }}
                enableEdit={true}
                enableComplete={true}
                store={store}
            />
        );

        wrapper = wrapperContainer.shallow();
    });

    afterEach(() => {
        store.clearActions();
    });

    it('should render correct with complete and edit features are enabled', () => {
        wrapper.setProps({
            enableEdit: true,
            enableComplete: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch completeTodo', () => {
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(0).find(ConfirmButton).prop('config').onOk();
        expect(store.getActions()).toEqual([completeTodo('1')]);
    });

    it('should show editable modal', () => {
        let show = jest.spyOn(TodoEditableModal, 'show');
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(1).find('a').at(0).simulate('click');
        expect(show.mock.calls.length).toEqual(1);
    });

    it('should dispatch editTodo', () => {
        let todoState = {
            id: '2',
            dueDatetime: 10,
            description: 'Hello',
            complete: false,
            createTime: 0,
        };
        let show = jest.spyOn(TodoEditableModal, 'show');
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(1).find('a').at(0).simulate('click');
        show.mock.calls[0][0].onSave(todoState);
        expect(store.getActions()).toEqual([editTodo(todoState)]);
    });

    it('should execute ipc sender method', () => {
        let dispatch = jest.spyOn(stickerDispatcher, 'dispatch');
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(2).find('a').at(0).simulate('click');
        expect(dispatch.mock.calls.length).toEqual(1);
        expect(dispatch.mock.calls[0][0]).toEqual({
            channel: 'createSticker',
            id: '1',
            description: 'Hello!',
        });
    });

    it('should dispatch deleteTodo', () => {
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(3).find(ConfirmButton).prop('config').onOk();
        expect(store.getActions()).toEqual([deleteTodo('1')]);
    });

    it('should render correct with complete feature is enabled', () => {
        wrapper.setProps({
            enableEdit: false,
            enableComplete: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct with edit feature is enabled', () => {
        wrapper.setProps({
            enableEdit: true,
            enableComplete: false,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct with complete and edit features are disabled', () => {
        wrapper.setProps({
            enableEdit: false,
            enableComplete: false,
        });
        expect(wrapper).toMatchSnapshot();
    });

});