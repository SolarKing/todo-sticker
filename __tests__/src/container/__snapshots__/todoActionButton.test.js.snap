// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoActionButton should render correct with complete and edit features are disabled 1`] = `
<div
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Delete",
                "okType": "danger",
                "onOk": [Function],
                "title": "Are you sure delete this todo?",
              }
            }
          >
            Delete
          </ConfirmButton>
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
</div>
`;

exports[`TodoActionButton should render correct with complete and edit features are enabled 1`] = `
<div
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Complete",
                "onOk": [Function],
                "title": "Are you sure complete this todo?",
              }
            }
          >
            Complete
          </ConfirmButton>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Edit
          </a>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Delete",
                "okType": "danger",
                "onOk": [Function],
                "title": "Are you sure delete this todo?",
              }
            }
          >
            Delete
          </ConfirmButton>
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
</div>
`;

exports[`TodoActionButton should render correct with complete feature is enabled 1`] = `
<div
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Complete",
                "onOk": [Function],
                "title": "Are you sure complete this todo?",
              }
            }
          >
            Complete
          </ConfirmButton>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Delete",
                "okType": "danger",
                "onOk": [Function],
                "title": "Are you sure delete this todo?",
              }
            }
          >
            Delete
          </ConfirmButton>
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
</div>
`;

exports[`TodoActionButton should render correct with edit feature is enabled 1`] = `
<div
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Edit
          </a>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <ConfirmButton
            config={
              Object {
                "cancelText": "No",
                "okText": "Delete",
                "okType": "danger",
                "onOk": [Function],
                "title": "Are you sure delete this todo?",
              }
            }
          >
            Delete
          </ConfirmButton>
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
</div>
`;
